geomesa-convert/geomesa-convert-common/src/main/scala/org/locationtech/geomesa/convert/SimpleFeatureValidator.scala

```diff
@@ -20,154 +21,155 @@ trait SimpleFeatureValidator {
-  private var lastErr: String = _
-
-  override def name: String = "z-index"
-
-  override def init(sft: SimpleFeatureType): Unit = {
-    valid = {
-      import org.locationtech.geomesa.utils.geotools.Conversions._
-      import org.locationtech.geomesa.utils.geotools.RichSimpleFeatureType.RichSimpleFeatureType
-
-      val maxDate = Date.from(BinnedTime.maxDate(sft.getZ3Interval).toInstant)
-      val minDate = Date.from(BinnedTime.ZMinDate.toInstant)
-      val dtgFn: (SimpleFeature) => Boolean = sft.getDtgIndex match {
-        case Some(dtgIdx) => (sf: SimpleFeature) =>
-          lastErr = null
-          if (sf.getAttribute(dtgIdx) == null){
-            lastErr = "Null date attribute"
-            false
-          } else if (sf.get[Date](dtgIdx).before(minDate)) {
-            lastErr = "Date is before Z3 Min Date"
-            false
-          } else if (sf.get[Date](dtgIdx).after(maxDate)) {
-            lastErr = "Date is after Z3 Max Date"
-            false
-          } else {
-            true
-          }
-        case None =>
-          throw new IllegalArgumentException("Z3 validator cannot be used on a type lacking a dtg index")
+      if (dtgIndex == -1 || geomIndex == -1) {
+        throw new IllegalArgumentException("Z3 validator cannot be used on a type lacking a date or geometry")
       }
+    }
 
-      val geomFn: (SimpleFeature) => Boolean =
-        if (sft.getGeometryDescriptor != null) {
-          (sf: SimpleFeature) =>
-            lastErr = null
-            if (sf.getDefaultGeometry == null) {
-              lastErr = "Geom is null"
-              false
-            } else if (!org.locationtech.geomesa.utils.geotools.wholeWorldEnvelope.contains(sf.geometry.getEnvelopeInternal)) {
-              lastErr = "Geom is not contained in whole world envelope"
-              false
-            } else {
-              true
-            }
-        } else {
-          throw new IllegalArgumentException("Z3 validator cannot be used on a type lacking a geom index")
-        }
-
-      (sf: SimpleFeature) => dtgFn(sf) && geomFn(sf)
+    override def validate(sf: SimpleFeature): Boolean = {
+      lastErr = null
+      val date = sf.getAttribute(dtgIndex).asInstanceOf[Date]
+      val geom = sf.getAttribute(geomIndex).asInstanceOf[Geometry]
+      if (date == null) {
+        lastErr = "Null date attribute"
+        false
+      } else if (geom == null) {
+        lastErr = "Null geometry attribute"
+        false
+      } else if (date.before(minDate)) {
+        lastErr = s"Date is before Z3 Min Date ($minDate)"
+        false
+      } else if (date.after(maxDate)) {
+        lastErr = s"Date is after Z3 Max Date ($maxDate)"
+        false
+      } else if (!wholeWorldEnvelope.contains(geom.getEnvelopeInternal)) {
+        lastErr = s"Geometry exceeds world bounds ($wholeWorldEnvelope)"
+        false
+      } else {
+        true
+      }
     }
+
+    override def lastError: String = lastErr
```

Comment:
[interesting] This line kind of freaks me out, even though there's nothing wrong with it.  `lastErr` is a mutable variable, but returning it from a method does not (evidently) pass the reference, because changing the value after you call this method does not change the value inside the class instance.  Weird.  Cool.

Reply:
yeah, I believe that strings and primitives are always passed by value, while other objects are passed by reference.